### DISCUSSION TITLE
Table Schema vs JSON Schema - what are the differences and which is preferable

### DISCUSSION THREAD

**Discussion by rufuspollock** [frictionlessdata/frictionlessdata.io#614] [Q&A] [answered]
@dafeder  asked:

> @rufuspollock we're also debating strategies for data dictionaries and data validation. We're obviously looking at table schema but also just describing tables with pure JSON Schema, just defining the rows as objects with specific properties. Are there any arguments for or against this approach you may have documented on your way to defining the table schema?

It's an excellent question and one we've thought about quite a bit (table schema and json schema)

And when we started json schema wasn't there. And we ourselves obviously use JSON Schema to define the data 

The shortest answer would be: if you are doing pure tables Table Schema is probably better. If you have data with nested structures then JSON Schema is better.

## Table Schema > JSON Schema

The main pros for table schema vs JSON schema is things like

* foreign keys
* some of the types which are more data oriented etc.
* basically if you are trying to map e.g. a postgres table, table schema may be a bit nicer.

But it's maybe more about the tooling than the schema per se e.g. for Table Schema you ahve goodtables (python and js etc) which is much more oriented to validating tabular data than e.g. json schema validators in my experience ... (e.g. goodtables has skip_rows etc etc).

All the kind of stuff you have when working with data (csv, etc) that is less relevant to json data.

Also i would say that json schema is now somewhat more complex. you have constraints e.g. oneof etc that are relevant for json schema but less so for tabular data (and which imply both engineering and performance implications).

## JSON Schema > Table Schema

* Some more complex validation options e.g. oneOf
* Supports full JSON i.e. nested structures not just flat tabular
* Good set of tooling

**Comment by roll** [comment, marked as answer]
@lwinfree @rufuspollock 
There is both question and answer in the same post so it cannot be marked as an answer. I think we just need to split it. 

[THE ANSWER](https://github.com/frictionlessdata/project/discussions/614)
